Preserve sandbox annotations for handling OCI hooks

saschagrunert opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
Some OCI hooks like the oci-seccomp-bpf-hook may re-use the sandbox
annotations which would get lost before applying this patch. This means
we now preserve all sandbox annotations for containers and overwrite
specific values during setup.

#### Which issue(s) this PR fixes:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->
None
#### Special notes for your reviewer:
None
#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
- Fixed bug that all custom sandbox annotations will be passed to OCI hooks and therefore are also available on the containers
```

haircommander commented:
I am pretty sure this could break kata, as there is a specific annotation passed to kata that specifies whether it's an infra container or not. PTAL @fidencio @amshinde

saschagrunert commented:
> I am pretty sure this could break kata, as there is a specific annotation passed to kata that specifies whether it's an infra container or not. PTAL @fidencio @amshinde

I'd fear that something like this happens… :scream:

saschagrunert commented:
/retest

saschagrunert commented:
We could build something with reflection to check for unknown fields and sync them up, otherwise I’m not sure how to solve the issue appropriately.

haircommander commented:
theoretically, if we started the annotations based on the sandbox ones, and applies all the rest on top, it would override any container overrides, but leave the sandbox ones behind. maybe that could work?

fidencio commented:
> I am pretty sure this could break kata, as there is a specific annotation passed to kata that specifies whether it's an infra container or not. PTAL @fidencio @amshinde

@haircommander, @saschagrunert,

I've tested this one with the following reproducer ...
```
[root@k8s cri-o]# cat ~/reproducer.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: fio-test
  labels:
    role: fio-test
spec:
  containers:
    - name: fio
      image: dmonakhov/alpine-fio
      command: ["fio", "--name=random-writers-emptydir", "--filename=/test-volume/testfile", "--ioengine=libaio", "--iodepth=16", "--rw=randwrite", "--bs=512", "--direct=0", "--size=1G", "--numjobs=1", "--runtime=60s", "--group_reporting", "--fallocate=none", "--time_based", "--thread", "--invalidate=1", "--output-format=json"]
      volumeMounts:
        - name: test-volume
          mountPath: /test-volume
  initContainers:
    - name: fedora
      image: fedora
      command: ["date"]
    - name: centos
      image: centos
      command: ["date"]
  restartPolicy: Never
  runtimeClassName: kata
  volumes:
    - name: test-volume
      emptyDir: {}
```

... and I was able to spawn the pod without issues.

Let me know if there's any other scenario, maybe a more complex one, that I could try.

saschagrunert commented:
Thanks for the test @fidencio! :pray: @haircommander what do we wanna do now?

except…
/retest

saschagrunert commented:
/retest

TomSweeneyRedHat commented:
@saschagrunert codecov is complaining that it wants more coverage.

umohnani8 commented:
Tests are passing and changes LGTM.
@saschagrunert @haircommander @fidencio are we good to get this in?

fidencio commented:
@umohnani8. from my side, yes! It didn't break kata-containers on a basic test I ran.

haircommander commented:
actually
/hold

I just realized we added annotations handling in pkg/container/ SpecAddAnnotations, can you maybe move this there? sorry for just remembering

saschagrunert commented:
> actually
> /hold
> 
> I just realized we added annotations handling in pkg/container/ SpecAddAnnotations, can you maybe move this there? sorry for just remembering

This comment has been slipped off… Rebased and addressed your comment. :+1:

saschagrunert commented:
/retest

saschagrunert commented:
/retest

saschagrunert commented:
/test e2e_features_rhel

haircommander commented:
/lgtm

haircommander commented:
none other than I forgot about it :)

saschagrunert commented:
/unhold

saschagrunert commented:
/retest
